src/mscorlib/shared/System/Threading/ExecutionContext.cs

```diff
@@ -130,126 +108,164 @@ public static bool IsFlowSuppressed()
 
         public static void Run(ExecutionContext executionContext, ContextCallback callback, Object state)
         {
+            // Note: ExecutionContext.Run is an extremly hot function and used by every await, threadpool execution etc
             if (executionContext == null)
+            {
                 throw new InvalidOperationException(SR.InvalidOperation_NullContext);
+            }
 
             Thread currentThread = Thread.CurrentThread;
-            ExecutionContextSwitcher ecsw = default(ExecutionContextSwitcher);
+            // Capture references to Thread Contexts
+            ref ExecutionContext current = ref currentThread.ExecutionContext;
+            ref SynchronizationContext currentSyncCtx = ref currentThread.SynchronizationContext;
+
+            // Store current ExecutionContext and SynchronizationContext as "previous"
+            // This allows us to restore them and undo any Context changes made in callback.Invoke
+            // so that they won't "leak" back into caller.
+            ExecutionContext previous = current;
+            SynchronizationContext previousSyncCtx = currentSyncCtx;
+
+            if (executionContext == Default)
+            {
+                // Default is a null ExecutionContext internally
+                executionContext = null;
+            }
+
+            if (previous != executionContext)
+            {
+                // Restore changed ExecutionContext
+                Restore(ref current, executionContext);
+            }
+
+            ExceptionDispatchInfo edi = null;
             try
             {
-                EstablishCopyOnWriteScope(currentThread, ref ecsw);
-                ExecutionContext.Restore(currentThread, executionContext);
-                callback(state);
+                callback.Invoke(state);
             }
-            catch
+            catch (Exception ex)
             {
                 // Note: we have a "catch" rather than a "finally" because we want
                 // to stop the first pass of EH here.  That way we can restore the previous
-                // context before any of our callers' EH filters run.  That means we need to
-                // end the scope separately in the non-exceptional case below.
-                ecsw.Undo(currentThread);
-                throw;
+                // context before any of our callers' EH filters run.
+                edi = ExceptionDispatchInfo.Capture(ex);
```

Comment:
Isn't this going to change the call stack from what it would have been, adding in the whole "rethrown from" gook?  That's going to affect pretty much every async invocation that incurs an exception, no?

Reply:
Its the same in both cases

Reply:
Really?  This:
```C#
using System;
using System.Runtime.ExceptionServices;

class Program
{
    static void Main()
    {
        try { Foo1(); } catch (Exception e) { Console.WriteLine(e.StackTrace); }
        Console.WriteLine();
        try { Foo2(); } catch (Exception e) { Console.WriteLine(e.StackTrace); }
    }

    static void Foo1()
    {
        try { throw new Exception("test"); }
        catch { throw; }
    }

    static void Foo2()
    {
        ExceptionDispatchInfo edi = null;
        try { throw new Exception("test"); }
        catch (Exception e) { edi = ExceptionDispatchInfo.Capture(e); }
        edi.Throw();
    }
}
```
produces this for me:
```
>corerun test.exe
   at Program.Foo1() in c:\Users\stoub\Desktop\CoreClrTest\test.cs:line 16
   at Program.Main() in c:\Users\stoub\Desktop\CoreClrTest\test.cs:line 8

   at Program.Foo2() in c:\Users\stoub\Desktop\CoreClrTest\test.cs:line 22
--- End of stack trace from previous location where exception was thrown ---
   at Program.Main() in c:\Users\stoub\Desktop\CoreClrTest\test.cs:line 10
```
You're saying though that with your change that ```--- End of stack trace...``` doesn't show up when an exception emerges from EC.Run?

Reply:
Been trying to get this to show up in async without much luck; but have since realized that async never let's an exception get caught here (i.e. its caught it to fail the Task)

Reply:
Added issue of EDI https://github.com/dotnet/coreclr/issues/15779

Reply:
Just need to change StackTrace from
```csharp
if (sf.GetIsLastFrameFromForeignExceptionStackTrace() &&
    !isAsync) // Skip EDI boundary for async
```
to
```csharp
if (sf.GetIsLastFrameFromForeignExceptionStackTrace() &&
    !isAsync && // Skip EDI boundary for async
    declaringType != typeof(ExecutionContext)) // Skip EDI boundary for ExecutionContext.Run
```
However would need to rebase; and also I said I'd finished making changes to this PR 😉